validation/validate.ts

```diff
@@ -0,0 +1,265 @@
+      // Any node that has an ancestor in nodeRefs is considered redundant because
+      // outerHTML includes all descendants.
+      const isRedundant = await page.evaluate(
+        (nodeA, nodeB) => nodeA.contains(nodeB),
+        nodeRefA.handle,
+        nodeRefB.handle
+      );
+      // nodeA contains nodeA, so we must ensure that nodeRefA !== nodeRefB or all
+      // nodes will be considered redundant.
+      if (isRedundant && nodeRefA !== nodeRefB) {
+        redundantNodes.push(nodeRefB);
+      }
+    }
+  }
+  return nodeRefs.filter(nodeRef => !redundantNodes.includes(nodeRef));
+}
+
+/**
+ * Gets all nodes used by Chrome to compute the accessible name for nodeRef.
+ * @param nodeRef - Node whose accessible name is being computed.
+ * @param client - CDPSession for page.
+ * @param page - Page containing nodeRef.
+ */
+async function getNodesUsedByChrome(
+  nodeRef: NodeRef,
+  client: CDPSession,
+  page: Page
+): Promise<NodeRef[]> {
+  const stack: NodeRef[] = [];
+  const nodesUsed: NodeRef[] = [];
+  // Track backendIds of visited nodes to avoid infinite cycle.
+  const visitedNodes: Protocol.DOM.BackendNodeId[] = [];
+  stack.push(nodeRef);
+  // Iterative DFS traverses nodes connected by label references
+  while (stack.length > 0) {
+    const currentNodeRef = stack.pop();
+    nodesUsed.push(currentNodeRef!);
+
+    const axTree = (await client.send('Accessibility.getPartialAXTree', {
+      backendNodeId: currentNodeRef!.backendId,
+    })) as Protocol.Accessibility.GetPartialAXTreeResponse;
+
+    // Check if any AXNodes descandant of currentNodeRef are labelled
+    for (const axNode of axTree.nodes) {
+      let labelNodes: Protocol.Accessibility.AXRelatedNode[] = [];
+      const sources: Protocol.Accessibility.AXValueSource[] =
+        axNode.name?.sources ?? [];
+
+      for (const source of sources) {
+        if (source.type === 'relatedElement') {
+          // Handles nodes connected by attribute value (aria-labelleby)
+          if (source.attributeValue?.relatedNodes) {
+            labelNodes = source.attributeValue.relatedNodes;
+            // Handles nodes connected natively (<label>)
+          } else if (source.nativeSourceValue?.relatedNodes) {
+            labelNodes = source.nativeSourceValue.relatedNodes;
+          }
+        }
+      }
+
+      // Repeat the process for all unvisited label nodes.
+      for (const labelNode of labelNodes) {
+        const labelNodeRef = await getNodeRefFromBackendId(
+          labelNode.backendDOMNodeId,
+          client,
+          page
+        );
+        if (labelNodeRef && !visitedNodes.includes(labelNodeRef.backendId)) {
+          stack.push(labelNodeRef);
+          visitedNodes.push(labelNodeRef.backendId);
+        }
+      }
+      // Stop iterating when we reach the current node :
+      // we have checked all descendants.
+      if (axNode.backendDOMNodeId === currentNodeRef!.backendId) {
+        break;
+      }
```

Comment:
I was quite confused reading this loop as `axTree.nodes` contains siblings and ancestors too so I was wondering why you were checking those nodes. I think it would help to move this before the loop so it's clear what's being iterated, by finding the index with `currentNodeRef.backendId` and slicing the array before iterating

Reply:
The reason I had this check at the end of the loop was to make sure that `currentNodeRef` was processed before breaking out of the loop, but I really like the suggestion of slicing the array before iterating -- implementing that now, thanks!